src/sentry/workflow_engine/models/detector.py

```diff
@@ -35,3 +51,63 @@ class Meta(OwnerModel.Meta):
                 name="workflow_engine_detector_org_name",
             )
         ]
+
+    @property
+    def detector_handler(self) -> DetectorHandler | None:
+        group_type = grouptype.registry.get_by_slug(self.type)
+        if not group_type:
+            logger.error(
+                "No registered grouptype for detector",
+                extra={
+                    "group_type": str(group_type),
+                    "detector_id": self.id,
+                    "detector_type": self.type,
+                },
+            )
+            return None
+
+        if not group_type.detector_handler:
+            logger.error(
+                "Registered grouptype for detector has no detector_handler",
+                extra={
+                    "group_type": str(group_type),
+                    "detector_id": self.id,
+                    "detector_type": self.type,
+                },
+            )
+            return None
+        return group_type.detector_handler(self)
+
+
+@dataclasses.dataclass(frozen=True)
+class DetectorStateData:
+    group_key: str | None
+    active: bool
+    status: DetectorStatus
+    # Stateful detectors always process data packets in order. Once we confirm that a data packet has been fully
+    # processed and all workflows have been done, this value will be used by the stateful detector to prevent
+    # reprocessing
+    dedupe_value: int
+    # Stateful detectors allow various counts to be tracked. We need to update these after we process workflows, so
+    # include the updates in the state
+    counter_updates: dict[str, int]
+
+
+@dataclasses.dataclass(frozen=True)
+class DetectorEvaluationResult:
+    is_active: bool
+    priority: PriorityLevel
+    data: Any
+    state_update_data: DetectorStateData | None = None
+
+
+T = TypeVar("T")
+
+
+class DetectorHandler(abc.ABC, Generic[T]):
+    def __init__(self, detector: Detector):
+        self.detector = detector
+
+    @abc.abstractmethod
+    def evaluate(self, data_packet: DataPacket[T]) -> list[DetectorEvaluationResult]:
```

Comment:
🤔 is there anything we can add to simplify writing the `evaluate` method per handler? it seems like each one will have to know about the data condition groups, how to evaluate conditions, and return results. maybe this could include some helper methods?

Reply:
We might need a `StatelessDetectorHandler` that can be used generically for most issue types. I think that as we move into concrete implementation we can see what things will be reused.

Probably all stateless issue types will use the same handler though.